samza-test/src/test/java/org/apache/samza/test/processor/TestZkStreamProcessorSession.java

```diff
@@ -0,0 +1,133 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.test.processor;
+
+import java.util.concurrent.CountDownLatch;
+import org.apache.samza.config.ZkConfig;
+import org.apache.samza.processor.StreamProcessor;
+import org.apache.samza.processor.TestZkStreamProcessorBase;
+import org.apache.samza.zk.ZkJobCoordinator;
+import org.junit.Assert;
+import org.junit.Test;
+
+
+/**
+ * Happy path tests.
+ * Start 1, 2, 5 processors and make sure they all consume all the events.
+ */
+public class TestZkStreamProcessorSession extends TestZkStreamProcessorBase {
+
+  @Override
+  protected String prefix() {
+    return "test_ZKS_";
+  }
+
+  @Test
+  public void testSingleStreamProcessor() {
+    testStreamProcessorWithSessionRestart(new String[]{"1"});
+  }
+
+  @Test
+  public void testTwoStreamProcessors() {
+    testStreamProcessorWithSessionRestart(new String[]{"2", "3"});
+  }
+
+  @Test
+  public void testFiveStreamProcessors() {
+    testStreamProcessorWithSessionRestart(new String[]{"4", "5", "6", "7", "8"});
+  }
+
+  private void testStreamProcessorWithSessionRestart(String[] processorIds) {
+
+    // set shorter session expiration for the test
+    map.put(ZkConfig.ZK_SESSION_TIMEOUT_MS, "500");
+
+    // create a latch of the size equals to the number of messages
+    int totalEventsToGenerate = 3 * messageCount;
+    TestZkStreamProcessorBase.TestStreamTask.endLatch = new CountDownLatch(totalEventsToGenerate);
+
+    // initialize the processors
+    StreamProcessor[] streamProcessors = new StreamProcessor[processorIds.length];
+    ZkJobCoordinator[] jobCoordinators = new ZkJobCoordinator[processorIds.length];
+    // we need to know when the processor has started
+    Object[] startWait = new Object[processorIds.length];
+    Object[] stopWait = new Object[processorIds.length];
+    for (int i = 0; i < processorIds.length; i++) {
+      startWait[i] = new Object();
+      stopWait[i] = new Object();
+      streamProcessors[i] = createStreamProcessor(processorIds[i], map, startWait[i], stopWait[i]);
+      jobCoordinators[i] = (ZkJobCoordinator) streamProcessors[i].getCurrentJobCoordinator();
+    }
+
+    // produce messageCount messages, starting with key 0
+    produceMessages(0, inputTopic, messageCount);
+
+    // run the processors in separate threads
+    Thread[] threads = new Thread[processorIds.length];
+    for (int i = 0; i < processorIds.length; i++) {
+      threads[i] = runInThread(streamProcessors[i], TestZkStreamProcessorBase.TestStreamTask.endLatch);
+      threads[i].start();
+      // wait until the processor reports that it has started
+      try {
+        synchronized (startWait[i]) {
+          startWait[i].wait(1000);
```

Comment:
This could wake up early due to a spurious wakeup. See Object#wait for details. You could address this either by adding a condition check or using a CountDownLatch, with the latter being simpler.